shared-module/audiodelays/PitchShift.c

```diff
@@ -0,0 +1,358 @@
+        m_malloc_fail(self->buffer_len);
+    }
+    memset(self->buffer[0], 0, self->buffer_len);
+
+    self->buffer[1] = m_malloc(self->buffer_len);
+    if (self->buffer[1] == NULL) {
+        common_hal_audiodelays_pitch_shift_deinit(self);
+        m_malloc_fail(self->buffer_len);
+    }
+    memset(self->buffer[1], 0, self->buffer_len);
+
+    self->last_buf_idx = 1; // Which buffer to use first, toggle between 0 and 1
+
+    // Initialize other values most effects will need.
+    self->sample = NULL; // The current playing sample
+    self->sample_remaining_buffer = NULL; // Pointer to the start of the sample buffer we have not played
+    self->sample_buffer_length = 0; // How many samples do we have left to play (these may be 16 bit!)
+    self->loop = false; // When the sample is done do we loop to the start again or stop (e.g. in a wav file)
+    self->more_data = false; // Is there still more data to read from the sample or did we finish
+
+    // The below section sets up the effect's starting values.
+
+    synthio_block_assign_slot(semitones, &self->semitones, MP_QSTR_semitones);
+    synthio_block_assign_slot(mix, &self->mix, MP_QSTR_mix);
+
+    // Allocate the window buffer
+    self->window_len = window; // bytes
+    self->window_buffer = m_malloc(self->window_len);
+    if (self->window_buffer == NULL) {
+        common_hal_audiodelays_pitch_shift_deinit(self);
+        m_malloc_fail(self->window_len);
+    }
+    memset(self->window_buffer, 0, self->window_len);
+
+    // Allocate the overlap buffer
+    self->overlap_len = overlap; // bytes
+    if (self->overlap_len) {
+        self->overlap_buffer = m_malloc(self->overlap_len);
+        if (self->overlap_buffer == NULL) {
+            common_hal_audiodelays_pitch_shift_deinit(self);
+            m_malloc_fail(self->overlap_len);
+        }
+        memset(self->overlap_buffer, 0, self->overlap_len);
+    } else {
+        self->overlap_buffer = NULL;
+    }
+
+    // The current position that the end of the overlap buffer will be written to the window buffer
+    self->window_index = 0;
+
+    // The position that the current sample will be written to the overlap buffer
+    self->overlap_index = 0;
+
+    // The position that the window buffer will be read from and written to the output
+    self->read_index = 0;
+
+    // Calculate the rate to increment the read index
+    mp_float_t f_semitones = synthio_block_slot_get(&self->semitones);
+    recalculate_rate(self, f_semitones);
+}
+
+bool common_hal_audiodelays_pitch_shift_deinited(audiodelays_pitch_shift_obj_t *self) {
+    if (self->window_buffer == NULL) {
+        return true;
+    }
+    return false;
+}
+
+void common_hal_audiodelays_pitch_shift_deinit(audiodelays_pitch_shift_obj_t *self) {
+    if (common_hal_audiodelays_pitch_shift_deinited(self)) {
+        return;
+    }
+    self->window_buffer = NULL;
+    self->overlap_buffer = NULL;
+    self->buffer[0] = NULL;
+    self->buffer[1] = NULL;
```

Comment:
setting these fields should remain, because it allows GC to release the related storage for object allocation.